### PR TITLE
bayes opt demo

### PR DESCRIPTION
**Pull request by wang90063**
### What changes were proposed in this pull request?

Add bayes_opt_sg_llama2.py and bayes_opt_sg_llama2_entry.sh in examples/llama2. 

Add description of bayesian optimization strategy generation algorithm in examples/llama2/README.md

### Does this PR introduce any user-facing change?

This PR gives the llama2 demo for users to try the  bayesian optimization strategy generation function

### How was this patch tested?

The demo has been tested and the results are listed in the examples/llama2/README.md

### CODE REVIEW

File: atorch/examples/llama2/README.md [hunk 1]
```diff
@@ -423,4 +423,104 @@ sh ds_3d_llama2_entry.sh
...
+5  amp_native   NotChosen  {"data": 8, "tensor": 1}  module_replace   NotChosen
+6  amp_native       zero1  {"data": 8, "tensor": 1}  module_replace  checkpoint
+7  NotChosen        zero2  {"data": 8, "tensor": 1}  NotChosen         NotChosen
+```
+
+Corresponding Dryrun Throughput
+
+```
+[[ -0.        ]
+ [-11.28738554]
```
Comment (skydoorkai) [inline atorch/examples/llama2/README.md:516]: Why throughput is negative?

File: atorch/examples/llama2/README.md [hunk 2]
```diff
@@ -423,4 +423,104 @@ sh ds_3d_llama2_entry.sh
...
+        torch.optim.AdamW,
+        train_dataset,
+        loss_func=my_loss_func,
+        prepare_input=my_prepare_input,
+        model_input_format="unpack_dict",
+        optim_args={"lr": args.learning_rate},
+        optim_param_func=partial(optim_param_func, args=args),
+        dataloader_args=dataloader_args,
+        excluded=[],
+        included=[{"atorch_wrap_cls": ('LlamaDecoderLayer')}],
```
Comment (skydoorkai) [inline atorch/examples/llama2/README.md:477]: included is a list of opts that should be included. This is not opt.

File: atorch/examples/llama2/README.md (hunk 1 above)
Comment (wang90063) [reply atorch/examples/llama2/README.md:516]: The BO optimizer is trying to minimize the object. Maxmizing equals miniming minus.

It is indeed confusing. I will correct it in the readme.

File: atorch/examples/llama2/README.md (hunk 2 above)
Comment (wang90063) [reply atorch/examples/llama2/README.md]: Could we include some parameters of opt?

Comment (skydoorkai) [reply atorch/examples/llama2/README.md]: Included is for opts.
If model info is required, add/use analysis task to get it.

Comment (wang90063) [reply atorch/examples/llama2/README.md]: ok. I will delete it.

Comment (wang90063) [reply atorch/examples/llama2/README.md]: And I will double check whether existed analysis tasks can capture LlamaDeocderLayer

Comment (skydoorkai) [reply atorch/examples/llama2/README.md]: Need to add LlamaDecoderLayer in _analyse_transformer.

File: atorch/examples/llama2/README.md [hunk 3]
```diff
@@ -1,426 +1,526 @@
-# Llama2 Pretrain/Finetune
```
Comment (nash635) [inline atorch/examples/llama2/README.md]: https://github.com/intelligent-machine-learning/dlrover/blob/master/atorch/docs/auto_accelerate_api.md
bayes opt的工作，可以也在auto acc的介绍主页上也提一下，使用方法可以指向这里

还有可以考虑添加一下bayes opt在auto_accelerate 中应用的具体设计原理文档，之前对外宣传的ppt直接链接上面那个md里面也可以。

Comment (wang90063) [reply atorch/examples/llama2/README.md]: “bayes opt的工作，可以也在auto acc的介绍主页上也提一下，使用方法可以指向这里” 这个没问题的，我在fully automatic optimization后面引用一下就好

“还有可以考虑添加一下bayes opt在auto_accelerate 中应用的具体设计原理文档，之前对外宣传的ppt直接链接上面那个md里面也可以。“ 这个材料得花几天时间整理一下

File: atorch/examples/llama2/README.md [hunk 4]
```diff
@@ -423,4 +423,98 @@ sh ds_3d_llama2_entry.sh
...
\ No newline at end of file
+| DS 3D | /        |          64 |  2 |  2 | 16 |             8 |              64 |              8192 | 153.93 | 49.3 | 22695.7               |
+
+## Automatic Training Optimization 
+### Introductin
+If the users are not sure what the strategy to achieve the largest throughput, auto_accelerate is able to automatically searching the best strategy given models and hardware conditions. This function is based on Bayesian Optimization (BO)implemented by [HEBO](https://github.com/huawei-noah/HEBO) aiming to find the strategy with largest training throughput efficiently. BO is a machine learning technique used for optimizing black-box functions that are expensive to evaluate. Specifically, BO learn the mapping from strategies to throughput using the data from dryrun, which run few training steps. Moreover, BO recommends the potential high-throughput strategy to achieve the throughput from dryrun, and update the mapping iteratively. This iterative process continues until the desired optimization criteria are met or a predefined budget is exhausted.
```
Comment (skydoorkai) [inline atorch/examples/llama2/README.md:430]: what the strategy -> which strategy would achieve 

This function is based on Bayesian Optimization... -> This is achieved using Bayesian Optimization...

Comment (skydoorkai) [reply atorch/examples/llama2/README.md:430]: BO learns the mapping between strategies to throughputs ...

and updates the ....

File: atorch/examples/llama2/README.md [hunk 5]
```diff
@@ -423,4 +423,98 @@ sh ds_3d_llama2_entry.sh
...
+
+### Usage
+In this section,an example is presented to show how to automatically search the best strategy. Specifically, we utilize 8 A100 to train the LLAMA2 7B model on wikitext-103-raw-v1. The shell script to run the experiment is given as follows
+```bash
```
Comment (skydoorkai) [inline atorch/examples/llama2/README.md:434]: This is the script in bayes_opt_sg_llama2_entry.sh ? Then just say that : the example source code is [bayes_opt_sg_llama2.py](bayes_opt_sg_llama2.py) and you can use [bayes_opt_sg_llama2_entry.sh](bayes_opt_sg_llama2_entry.sh) to run the example.
Then explains the env variables and auto_acceleate API parameters.

**Review by skydoorkai — APPROVED** [review]
lgtm

**Review by nash635 — APPROVED** [review]

**Pull request merged**